File: src/main/java/com/example/member/service/MemberService.java
package com.example.member.service;

import com.example.member.domain.Member;
import com.example.member.repository.MemoryMemberRepository;
import com.example.member.repository.MemberRepository;

import java.util.List;
import java.util.Optional;

public class MemberService {
    private final MemberRepository memberRepository = new MemoryMemberRepository();
    
    public Long join(Member member) {
        validateDuplicateMember(member);

        memberRepository.save(member);
        return member.getId();
    }

    private void validateDuplicateMember(Member member) {
        memberRepository.findByName(member.getName())
        .ifPresent(m -> {
            throw new IllegalStateException("이미 존재하는 회원입니다.");
        });
    }
    
    public List<Member> findMembers() {
        return memberRepository.findAll();
    }
    
    public Optional<Member> findMember(Long memberId) {
        return memberRepository.findById(memberId);
    }
}

File: 스프링 부트 핵심 가이드/src/main/kotlin/dong/kotlin_study/service/MemberService.kt
package dong.kotlin_study.service

import dong.kotlin_study.domain.Member
import dong.kotlin_study.repository.MemberRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class MemberService(val memberRepository: MemberRepository) {
    fun join(member: Member): Long {
        validateDuplicateMember(member)
        memberRepository.save(member)
        return member.id!!
    }

    private fun validateDuplicateMember(member: Member) {
        memberRepository.findByName(member.name)
            .ifPresent { throw IllegalStateException("이미 존재하는 회원입니다.") }
    }

    fun findMembers(): List<Member> = memberRepository.findAll()
    fun findOne(memberId: Long): Optional<Member> = memberRepository.findById(memberId)
}